### PR TITLE
SDI-363 error handling

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Send back error codes with errors and add integration tests for error scenarios

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt [hunk 1]
```diff
@@ -15,7 +15,7 @@ import org.springframework.web.client.RestClientException
 import org.springframework.web.client.RestClientResponseException
 import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
 import org.springframework.web.reactive.function.client.WebClientResponseException
-import uk.gov.justice.digital.hmpps.hmppsrestrictedpatientsapi.model.exceptions.NoResultsReturnedException
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt:18]: Dropped this custom exception - it behaves exactly like t he EntityNotFoundException so using that instead.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt [hunk 2]
```diff
@@ -35,6 +35,21 @@ class HmppsRestrictedPatientsApiExceptionHandler {
...
+      .status(BAD_REQUEST)
+      .body(
+        ErrorResponse(
+          status = BAD_REQUEST,
+          errorCode = e.errorCode,
+          developerMessage = "Validation failure: ${e.message}",
+          userMessage = e.message
+        )
+      )
+  }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt:51]: Added the custom BadRequestException to hold an errorCode that we can send back to the caller helping them differentiate.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt [hunk 3]
```diff
@@ -78,6 +93,7 @@ class HmppsRestrictedPatientsApiExceptionHandler {
       .body(
         ErrorResponse(
           status = INTERNAL_SERVER_ERROR.value(),
+          errorCode = "UPSTREAM_ERROR",
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt:96]: Differentiates errors from failing upstream services with 500s caused by a bug in our service. Helps the client display nicer error messages to end users (e.g. whether to retry later).

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt [hunk 4]
```diff
@@ -190,17 +193,19 @@ class HmppsRestrictedPatientsApiExceptionHandler {
 
 data class ErrorResponse(
   val status: Int,
-  val errorCode: Int? = null,
+  val errorCode: String? = null,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/config/HmppsRestrictedPatientsApiExceptionHandler.kt:196]: OK to change as not currently being used. Much prefer a descriptive String to some random Int.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/repositories/RestrictedPatientsRepository.kt [hunk 5]
```diff
@@ -1,6 +1,8 @@
 package uk.gov.justice.digital.hmpps.hmppsrestrictedpatientsapi.repositories
 
 import org.springframework.data.jpa.repository.JpaRepository
+import org.springframework.stereotype.Repository
 import uk.gov.justice.digital.hmpps.hmppsrestrictedpatientsapi.model.entities.RestrictedPatient
 
+@Repository
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/repositories/RestrictedPatientsRepository.kt:7]: Stop Intellij warning.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/integration/wiremock/PrisonApiMockServer.kt [hunk 6]
```diff
@@ -276,4 +269,15 @@ class PrisonApiMockServer : WireMockServer(8989) {
         )
     )
   }
+
+  fun stubServerError(method: (urlPattern: UrlPattern) -> MappingBuilder) {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsrestrictedpatientsapi/integration/wiremock/PrisonApiMockServer.kt:273]: `method` would be a WireMock::get (or post etc.)

**Review by lbennett-moj — APPROVED** [review]

**Pull request merged**